File: iosApp/iosApp/Helpers/mappers/GameSettingsMapper.swift
import ComposeApp

extension IOSMultiplayerManager {
    func dictionary(from gameSettings: GameSettings) -> [String: Any] {
        [
            "isNotificationEnabled": gameSettings.isNotificationEnabled,
            "isVirtualDiceEnabled": gameSettings.isVirtualDiceEnabled,
            "isShakeEnabled": gameSettings.isShakeEnabled,
            "hasStartLimit": gameSettings.hasStartLimit,
            "isBarrel1Active": gameSettings.isBarrel1Active,
            "isBarrel2Active": gameSettings.isBarrel2Active,
            "isBarrel3Active": gameSettings.isBarrel3Active,
            "isTripleBoltFineActive": gameSettings.isTripleBoltFineActive,
            "isOvertakeFineActive": gameSettings.isOvertakeFineActive
        ]
    }

    func gameSettings(from value: Any?) -> GameSettings {
        guard let dictionary = value as? [String: Any] else {
            return defaultGameSettings()
        }

        return GameSettings(
            isNotificationEnabled: dictionary["isNotificationEnabled"] as? Bool ?? true,
            isVirtualDiceEnabled: dictionary["isVirtualDiceEnabled"] as? Bool ?? true,
            isShakeEnabled: dictionary["isShakeEnabled"] as? Bool ?? true,
            hasStartLimit: dictionary["hasStartLimit"] as? Bool ?? true,
            isBarrel1Active: dictionary["isBarrel1Active"] as? Bool ?? true,
            isBarrel2Active: dictionary["isBarrel2Active"] as? Bool ?? true,
            isBarrel3Active: dictionary["isBarrel3Active"] as? Bool ?? false,
            isTripleBoltFineActive: dictionary["isTripleBoltFineActive"] as? Bool ?? true,
            isOvertakeFineActive: dictionary["isOvertakeFineActive"] as? Bool ?? true
        )
    }

    func defaultGameSettings() -> GameSettings {
        GameSettings(
            isNotificationEnabled: true,
            isVirtualDiceEnabled: true,
            isShakeEnabled: true,
            hasStartLimit: true,
            isBarrel1Active: true,
            isBarrel2Active: true,
            isBarrel3Active: false,
            isTripleBoltFineActive: true,
            isOvertakeFineActive: true
        )
    }
}

File: composeApp/src/commonMain/kotlin/com/alextos/thousand/data/mappers/firebase/GameSettingsMapper.kt
package com.alextos.thousand.data.mappers.firebase

import com.alextos.thousand.domain.models.GameSettings

object FirebaseGameSettingsMapper {
    fun dictionary(from: GameSettings): Map<String, Any> =
        from.toFirebaseMap()

    fun gameSettings(from: Any?): GameSettings =
        from.asFirebaseMap().toFirebaseGameSettings()
}

internal fun GameSettings.toFirebaseMap(): Map<String, Any> =
    mapOf(
        "isNotificationEnabled" to isNotificationEnabled,
        "isVirtualDiceEnabled" to isVirtualDiceEnabled,
        "isShakeEnabled" to isShakeEnabled,
        "hasStartLimit" to hasStartLimit,
        "isBarrel1Active" to isBarrel1Active,
        "isBarrel2Active" to isBarrel2Active,
        "isBarrel3Active" to isBarrel3Active,
        "isTripleBoltFineActive" to isTripleBoltFineActive,
        "isOvertakeFineActive" to isOvertakeFineActive,
    )

internal fun Map<*, *>?.toFirebaseGameSettings(): GameSettings =
    GameSettings(
        isNotificationEnabled = boolean("isNotificationEnabled", true),
        isVirtualDiceEnabled = boolean("isVirtualDiceEnabled", true),
        isShakeEnabled = boolean("isShakeEnabled", true),
        hasStartLimit = boolean("hasStartLimit", true),
        isBarrel1Active = boolean("isBarrel1Active", true),
        isBarrel2Active = boolean("isBarrel2Active", true),
        isBarrel3Active = boolean("isBarrel3Active", false),
        isTripleBoltFineActive = boolean("isTripleBoltFineActive", true),
        isOvertakeFineActive = boolean("isOvertakeFineActive", true),
    )
